monitoring/nethealth.go

```diff
@@ -0,0 +1,411 @@
+type NethealthConfig struct {
+	// AdvertiseIP specifies the advertised ip address of the host running this checker.
+	AdvertiseIP string
+	// NethealthPort specifies the port that nethealth is listening on.
+	NethealthPort int
+	// NetStatsInterval specifies the duration to store net stats.
+	NetStatsInterval time.Duration
+	// KubeConfig specifies kubernetes access information.
+	*KubeConfig
+}
+
+// CheckAndSetDefaults validates that this configuration is correct and sets
+// value defaults where necessary.
+func (c *NethealthConfig) CheckAndSetDefaults() error {
+	var errors []error
+	if c.AdvertiseIP == "" {
+		errors = append(errors, trace.BadParameter("host advertise ip must be provided"))
+	}
+	if c.NethealthPort == 0 {
+		c.NethealthPort = defaultNethealthPort
+	}
+	if c.NetStatsInterval == time.Duration(0) {
+		c.NetStatsInterval = defaultNetStatsInterval
+	}
+	if c.KubeConfig == nil {
+		errors = append(errors, trace.BadParameter("kubernetes access config must be provided"))
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// nethealthChecker checks network communication between peers.
+type nethealthChecker struct {
+	// NethealthConfig contains caller specified nethealth checker configuration
+	// values.
+	NethealthConfig
+	// seriesCapacity specifies the number of data points to store in a series.
+	seriesCapacity int
+	// lock access to timeoutStats
+	sync.Mutex
+	// peerStats maps a peer to its recorded nethealth stats.
+	peerStats netStats
+}
+
+// NewNethealthChecker returns a new nethealth checker.
+func NewNethealthChecker(config NethealthConfig) (*nethealthChecker, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// seriesCapacity is determined by the NetStatsInterval / StatusUpdateTimeout
+	// rounded up to the nearest integer
+	seriesCapacity := math.Ceil(float64(config.NetStatsInterval) / float64(agent.StatusUpdateTimeout))
+
+	return &nethealthChecker{
+		peerStats:       newNetStats(netStatsCapacity),
+		seriesCapacity:  int(seriesCapacity),
+		NethealthConfig: config,
+	}, nil
+}
+
+// Name returns this checker name
+// Implements health.Checker
+func (c *nethealthChecker) Name() string {
+	return nethealthCheckerID
+}
+
+// Check verifies the network is healthy.
+// Implements health.Checker
+func (c *nethealthChecker) Check(ctx context.Context, reporter health.Reporter) {
+	probes, err := c.check(ctx)
+	if err != nil {
+		log.WithError(err).Error("Unable to verify nethealth.")
+	}
+
+	if probes.NumProbes() != 0 {
+		health.AddFrom(reporter, probes)
+		return
+	}
+	reporter.Add(NewSuccessProbe(c.Name()))
+}
+
+func (c *nethealthChecker) check(ctx context.Context) (health.Reporter, error) {
+	reporter := new(health.Probes)
+	pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		reporter.Add(NewProbeFromErr(c.Name(), "failed to retrieve k8s pods", err))
```

Comment:
And these will go away as we'll only be adding probes in the top-level Check method.